cmd/kubeadm/app/phases/upgrade/staticpods.go

```diff
@@ -305,6 +302,8 @@ func performEtcdStaticPodUpgrade(certsRenewMgr *renewal.Manager, client clientse
 	if err != nil {
 		return true, errors.Wrapf(err, "failed to parse the current etcd version(%s)", currentEtcdVersionStr)
 	}
+	// For version comparison, we will only use MAJOR.MINOR.PATCH
+	currentEtcdVersion = currentEtcdVersion.WithPreRelease("")
```

Comment:
> Also, the comparison between desired and current etcd versions is not
accurate, as the desired version might contain prerelease information,
whereas the current etcd version is retrieved from the running etcd
cluster runtime information. In order to make this comparison more
even, both versions will strip pre-release information, and only
MAJOR.MINOR.PATCH will be considered on this comparison.

can you provide examples?

a version with a pre-release is older than the version of without it.
`v1.17.0-alpha.1` is older than `v1.17.0`

so not sure if clamping one of the versions is valid here.

Reply:
We are clamping both when comparing. About the mismatch explanation, see:

https://github.com/kubernetes/kubernetes/blob/67c6767b7da983034be04a31575261890186338a/cmd/kubeadm/app/constants/constants.go#L437

vs

```
root@kind-control-plane:/# etcdctl --cacert=/etc/kubernetes/pki/etcd/ca.crt --cert=/etc/kubernetes/pki/etcd/healthcheck-client.crt --key=/etc/kubernetes/pki/etcd/healthcheck-client.key version                                                                                                                                                      
etcdctl version: 3.4.3
API version: 3.4
```

Reply:
Let me fetch the real version returned in this case by the gRPC call.

Reply:
@neolit123 updated the comment above.

Reply:
just so that others can see this too, the `-0` suffix is the revision that Google wants to have at GCR. in the future this might happen for all CP images!

still trying to understand more here. excluding `--config` from the equation:
- if the cluster version is `3.4.3`, but kubeadm wants to upgrade to `3.5.0` this will be a non issue, correct?
- if the cluster version is `3.4.3`, but kubeadm want's to apply `3.4.3-0` is this branch entered?

```
if desiredEtcdVersion.LessThan(currentEtcdVersion) {
		return false, errors.Errorf("the desired etcd version for this Kubernetes version %q is %q, but the current etcd version is %q. Won't downgrade etcd, instead just continue", cfg.KubernetesVersion, desiredEtcdVersion.String(), currentEtcdVersion.String())
	}
```

if yes, clamping here seems valid, but that does feel like a separate commit.

> This patch changes the logic, so the desired etcd version is retrieved
from the static map only if we are fetching the configuration from the
cluster. Otherwise, honor the etcd image tag provided in the
configuration.

my understanding here is that in your setup you are using `--config` (which we all know is discouraged) to reconfigure the etcd image _during_ upgrade.

can you share the image tag you are using?

also can you please provide the exact story of the bug:
1. kubeadm receives config with etcd version Y
1. kubeadm tries to upgrade to k8s version X
1. ...
...

thanks

Reply:
/cc @rosti

Reply:
I mostly agree with @neolit123 here. Another possibility is that at some point we might want to upgrade etcd `3.4.3-21` to `3.4.3-42` or we have used an RC release for some reason and want to move to the stable one.
I too think, that this is a separate thing and needs a separate discussion. It's kind of controversial to me too.
/hold

Reply:
+1, I will strip that part from this PR so we can discuss that separately.

Reply:
> my understanding here is that in your setup you are using --config (which we all know is discouraged) to reconfigure the etcd image during upgrade.
>
> can you share the image tag you are using?

We are using a different tagging strategy (e.g. `3.4.1-rev2`).

> also can you please provide the exact story of the bug:

You want to upgrade an existing cluster with `kubeadm upgrade apply`. Since we are not using the inernal `kubeadm` container version enforcement, we generate a kubeadm configuration file with the exact versions that we want of every component. Then, we run `kubeadm upgrade apply --config ...` and `kubeadm upgrade node` afterwards on every node.

The main idea here, and the rationale behind this change, is that everything on the provided configuration file is honored when calling to `kubeadm upgrade apply --config`, except for the etcd version, that is hardcoded and overriden always based on the `kubernetesVersion` provided. If a fully fledged configuration file is passed to `kubeadm upgrade apply` with `--config`, then I believe `etcd` should as well be honored as was specified in the configuration.

Reply:
> Another possibility is that at some point we might want to upgrade etcd 3.4.3-21 to 3.4.3-42 or we have used an RC release for some reason and want to move to the stable one.

true, the clamps will trip us if we start having revision based increments.

Reply:
@ereslibre do you have plans to move away from `--config` for inplace upgrades?

the workaround is the following:
- let kubeadm upgrade to it's preferred etcd.
- patch the ClusterConfiguration and etcd.yaml manifests (or call the etcd phase to generate them).

Reply:
> @ereslibre do you have plans to move away from --config for inplace upgrades?

At the moment it would be desirable to honor the configuration pased with `--config`. Note that we will upgrade to the preferred `etcd`, only that our tags may not match exactly with the default ones present in gcr.io.

In our case, the first step would not be ideal, for several reasons:

- If our registry is used, that tag may not exist at all, we would be forced to maintain kubeadm's tagging.

- If we let the pull to happen from gcr.io, we would be breaking airgap use cases.

I think it's reasonable to either do not supply a config at all (in which case, kubeadm's behavior doesn't change), or if a config is used, etcd will be taken from the provided configuration.

Reply:
(github) please "unresolve" this discussion for better visibility.

> At the moment it would be desirable to honor the configuration pased with --config.

if we are to support --config long term for upgrade this needs a new format `UpgradeConfiguration`, passing ClusterConfiguration and or InitConfiguration can be messy.

> If our registry is used, that tag may not exist at all, we would be forced to maintain kubeadm's tagging.

can aliasing tags help here as a workaound?
e.g. if during upgrades kubeadm wants etcd `3.7.1-0` you can create a tag reference using e.g. `docker tag`

Reply:
> can aliasing tags help here as a workaound?

I fear this is not enough. Sometimes we will want to create a new tag with some specific fix that is not available upstream or tagged upstream, and we still want our customers to be able to upgrade to that version, using our registry. This is version handling is completely on us, and we are using `kubeadm` as the driving logic during the upgrade, but not necessarily for resolving what tags of what images we need.